### PR TITLE
[FRR][sonic-cfggen] Support T2 chassis frontend

### PR DESCRIPTION
**Pull request by baiwei0427**
**- What I did**

Support new role: T2 chassis frontend switch (SpineChassisFrontendRouter). The chassis frontend switch needs BGP-EVPN, Vxlan and Vnet support.
 
**- How I did it**

* Modify minigraph parser to add Vnet and Vxlan informration for T2 chassis frontend

* Add new frr, bgpd and zebra j2 templates to support BGP-EVPN

* Add new tests to sonic-cfggen unit test scripts

**- How to verify it**

Run load_minigraph and verify entries in config DB and FRR configuration.

In config DB, you should be able to see the following entries:

```
"VNET|VnetFE"
"VXLAN_TUNNEL|TunnelInt"
```

In FRR docker, you should be able to see BGP-EVPN configuration as follows:

```
hostname SpineFront01
password zebra
log syslog informational
log facility local4
! enable password !
! Vnet BGP instance
vrf VnetFE
vni 8000
router bgp 4000 vrf VnetFE
  no bgp default ipv4-unicast
  bgp log-neighbor-changes
  bgp bestpath as-path multipath-relax
  no bgp default ipv4-unicast
  bgp graceful-restart restart-time 240
  bgp graceful-restart
  bgp router-id 4.0.0.0
  neighbor 192.168.0.1 remote-as 3000
  neighbor 192.168.0.1 description Leaf01
  neighbor 192.168.0.1 timers 3 10
  address-family ipv4 unicast
    neighbor 192.168.0.1 activate
    neighbor 192.168.0.1 soft-reconfiguration inbound
    maximum-paths 64
  exit-address-family
  address-family l2vpn evpn
    advertise ipv4 unicast
  exit-address-family

!
! bgp multiple-instance
!
route-map FROM_BGP_SPEAKER_V4 permit 10
!
route-map TO_BGP_SPEAKER_V4 deny 10
!
router bgp 4000
  bgp log-neighbor-changes
  bgp bestpath as-path multipath-relax
  no bgp default ipv4-unicast
  bgp graceful-restart restart-time 240
  bgp graceful-restart
  bgp router-id 4.0.0.0
  network 4.0.0.0/32
  neighbor 4.0.0.1 remote-as 4000
  neighbor 4.0.0.1 description SpineFront02
  neighbor 4.0.0.1 timers 3 10
  address-family l2vpn evpn
    neighbor 4.0.0.1 activate
    advertise-all-vni
  exit-address-family
  neighbor 172.16.0.2 remote-as 5000
  neighbor 172.16.0.2 description SpineBack01
  neighbor 172.16.0.2 timers 3 10
  address-family ipv4 unicast
    neighbor 172.16.0.2 allowas-in 1
    neighbor 172.16.0.2 activate
    neighbor 172.16.0.2 soft-reconfiguration inbound
    maximum-paths 64
  exit-address-family
  neighbor 172.16.0.10 remote-as 5000
  neighbor 172.16.0.10 description SpineBack02
  neighbor 172.16.0.10 timers 3 10
  address-family ipv4 unicast
    neighbor 172.16.0.10 allowas-in 1
    neighbor 172.16.0.10 activate
    neighbor 172.16.0.10 soft-reconfiguration inbound
    maximum-paths 64
  exit-address-family
!
maximum-paths 64
!
route-map ISOLATE permit 10
set as-path prepend 4000
!

```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: Wei Bai <baiwei0427@gmail.com> and Ze Gan <ganze718@gmail.com>

### CODE REVIEW

File: src/sonic-config-engine/minigraph.py [hunk 1]
```diff
@@ -643,6 +643,52 @@ def parse_xml(filename, platform=None, port_config_file=None):
...
 
+ 
+    spine_chassis_frontend_role = 'SpineChassisFrontendRouter'
+    chassis_backend_role = 'ChassisBackendRouter'
+
+    # Special parsing for spine chassis frontend routers
+    if current_device['type'] == spine_chassis_frontend_role:
+        chassis_vnet ='Vnet1'
+        chassis_vxlan_tunnel = 'tunnel1'
+        chassis_vni = 8000
```
Comment (lguohan) [inline src/sonic-config-engine/minigraph.py]: I think we should not hard code here. We can hardcode in the template.

File: src/sonic-config-engine/minigraph.py [hunk 2]
```diff
@@ -643,6 +643,52 @@ def parse_xml(filename, platform=None, port_config_file=None):
...
+        lo_addr = '0.0.0.0'
+        for lo in lo_intfs:
+            lo_network = ipaddress.IPNetwork(lo[1])
+            if lo_network.version == 4:
+                lo_addr = str(lo_network.ip)
+                break        
+
+        results['VXLAN_TUNNEL'] = {chassis_vxlan_tunnel: {
+            'source_ip': lo_addr
+        }}
```
Comment (lguohan) [inline src/sonic-config-engine/minigraph.py]: can we use the lo_addr directly in the bgpd.conf j2 template?

File: src/sonic-config-engine/minigraph.py [hunk 3]
```diff
@@ -643,6 +643,52 @@ def parse_xml(filename, platform=None, port_config_file=None):
...
+            intf_name = intf[0]
+            neighbor_router = results['DEVICE_NEIGHBOR'][intf_name]['name']
+            
+            # if the neighbor router is an external router (neither spine frontend nor backend)
+            if devices[neighbor_router]['type'] != spine_chassis_frontend_role and \
+               devices[neighbor_router]['type'] != chassis_backend_role:
+
+                # Enslave the interface to a Vnet
+                if intf_name in phyport_intfs:
+                    phyport_intfs[intf_name] = {'vnet_name': chassis_vnet}
```
Comment (lguohan) [inline src/sonic-config-engine/minigraph.py]: should we move this enslave logic into the template as well?

File: src/sonic-config-engine/minigraph.py [hunk 4]
```diff
@@ -643,6 +643,52 @@ def parse_xml(filename, platform=None, port_config_file=None):
...
+                else:
+                    print >> sys.stderr, 'Warning: cannot find the key %s' % (intf_name) 
+
     return results
```
Comment (lguohan) [inline src/sonic-config-engine/minigraph.py]: all the extra logics you add in this file seem unnecessary.

File: src/sonic-config-engine/minigraph.py (hunk 1 above)
Comment (baiwei0427) [reply src/sonic-config-engine/minigraph.py]: This information (Vnet name, Vxlan tunnel name and vni) will be inserted to config DB. I don't know which template you refer to.

Comment (lguohan) [reply src/sonic-config-engine/minigraph.py]: bgpd.j2

Comment (baiwei0427) [reply src/sonic-config-engine/minigraph.py]: @lguohan I think this information is also necessary for fpmsyncd and vxlanmgrd

File: src/sonic-config-engine/minigraph.py [hunk 5]
```diff
@@ -643,6 +643,52 @@ def parse_xml(filename, platform=None, port_config_file=None):
             count += 1
         results['MIRROR_SESSION'] = mirror_sessions
 
+ 
+    spine_chassis_frontend_role = 'SpineChassisFrontendRouter'
+    chassis_backend_role = 'ChassisBackendRouter'
+
+    # Special parsing for spine chassis frontend routers
+    if current_device['type'] == spine_chassis_frontend_role:
+        chassis_vnet ='Vnet1'
```
Comment (lguohan) [inline src/sonic-config-engine/minigraph.py]: VnetFE

File: src/sonic-config-engine/minigraph.py [hunk 6]
```diff
@@ -643,6 +643,52 @@ def parse_xml(filename, platform=None, port_config_file=None):
...
+        chassis_vxlan_tunnel = 'tunnel1'
```
Comment (lguohan) [inline src/sonic-config-engine/minigraph.py]: TunnelInt

File: src/sonic-config-engine/minigraph.py [hunk 7]
```diff
@@ -643,6 +643,52 @@ def parse_xml(filename, platform=None, port_config_file=None):
...
+            'source_ip': lo_addr
+        }}
+
+        # Vnet information
+        results['VNET'] = {chassis_vnet: {
+            'vxlan_tunnel': chassis_vxlan_tunnel,
+            'vni': chassis_vni
+        }}
+
+        for intf in phyport_intfs:
```
Comment (lguohan) [inline src/sonic-config-engine/minigraph.py]: what if the front panel port channel belong to vnet?

**Comment by baiwei0427** [comment]
retest this please

**Comment by lguohan** [comment]
can you resolve conflict?

File: src/sonic-config-engine/tests/test_minigraph_case.py [hunk 8]
```diff
@@ -123,3 +125,39 @@ def test_metadata_ntp(self):
...
+        output = self.run_script(argument)
+        self.assertEqual(output.strip(), 'SpineChassisFrontendRouter')
+
+    def test_minigraph_t2_chassis_frontend_interfaces(self):
+        argument = '-m "' + self.sample_graph_t2_chassis_frontend + '" -p "' + self.t2_chassis_frontend_port_config + '" -v "INTERFACE"'
+        output = self.run_script(argument)
+        self.assertEqual(output.strip(),  
+                         "{'Ethernet8': {}, "
+                         "('Ethernet8', '172.16.0.9/30'): {}, "
+                         "'Ethernet0': {'vnet_name': 'VnetFE'}, "
```
Comment (lguohan) [inline src/sonic-config-engine/tests/test_minigraph_case.py]: need port channel example.

**Comment by baiwei0427** [comment]
New PR: https://github.com/Azure/sonic-buildimage/pull/3100

**Pull request closed without merging**